src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+            subshards = IntervalUtils.cutToShards(ints, outputShardSize);
+            currentSubShardIndex = 0;
+            currentSubShard = subshards.get(currentSubShardIndex);
+
+            if (BucketUtils.isCloudStorageUrl(bam)) {
+                storageClient = auth.makeStorageClient();
+                reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.SILENT);
+            } else if (BucketUtils.isHadoopUrl(bam)) {
+                throw new RuntimeException("Sorry, Hadoop paths aren't yet supported");
+            } else {
+                // read from local file (this only makes sense if every worker sees the same thing, e.g. if we're running locally)
+                reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
+            }
+            query = reader.queryOverlapping(shard.interval.getContig(), shard.interval.getStart(), shard.interval.getEnd());
+
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (null==nextOutput) nextOutput = tryNext();
+            return (null!=nextOutput);
+        }
+
+        @Override
+        public ContextShard next() {
+            if (null==nextOutput) nextOutput = tryNext();
+            if (null==nextOutput) throw new NoSuchElementException();
+            ContextShard ret =  nextOutput;
+            nextOutput = null;
+            return ret;
+        }
+
+        // returns the next shard, if any. Otherwise, closes the reader and returns null.
+        private ContextShard tryNext() {
+            if (readerClosed) return null;
+            ContextShard ret = null;
+            while (query.hasNext()) {
+
+                SAMRecord r = query.next();
+                SAMRecordToGATKReadAdapter g = new SAMRecordToGATKReadAdapter(r);
+                // yes, it'd be a tad faster to check before the wrapping.
+                // But this keeps the code a tad simpler.
+                if (!accept(g, shard.interval)) continue;
+                if (null!=optFilter) {
+                    // skip reads that don't pass the filter
+                    if (!optFilter.test(g)) continue;
+                }
+                if (!g.isUnmapped()) {
+                    // error out if we accept a read that sticks out too far
+                    // (margin was too tight, the shard may end up missing relevant variants)
+                    if (r.getAlignmentEnd()>lastValidPos) {
+                        throw new GATKException("Margin was too tight, a read sticks out by "+(r.getAlignmentEnd()-shard.interval.getEnd())+", going all the way to "+r.getAlignmentEnd());
+                    }
+                    if (r.getAlignmentStart()<firstValidPos) {
+                        throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                    }
+                }
```

Comment:
ok, except I'm choosing a different name because the name "assert" [suggests](http://docs.oracle.com/javase/specs/jls/se8/html/jls-14.html#jls-14.10) (i) it'll throw an AssertionError if violated, and (ii) it's disabled in production.